eclipse-jdt/eclipse.jdt.core
# Enhance resource leak analysis with additional annotation(s)

fixes #1530

## What it does

Demonstrate how evaluating one more annotations `@Owning` can enhance our resource leak analysis.

Most of the change is infra structure work to define new options, let the compiler detect that annotation and propagate that information to where it is needed.

Few semantically interesting code changes suffice to tell the existing analysis:
1. a method parameter of a closeable type and marked `@Owning` **must** be closed in, or on behalf of this method.
2. passing an unclosed closeable value into a method where the receiving parameter is marked `@Owning` delegates responsibility of closing to the receiving method.
3. returning an unclosing closeable value from a method annotated as `@Owning` delegates responsibility to the calling method.

The above can be mapped as equivalent to existing constructs:
1. corresponds to acquiring a fresh resource, e.g., by allocating a new instance.
2. corresponds to calling `close()` on the resource being sent to the other method
3. corresponds to calling `close()` on the resource being returned from the current method

## How to test

1. Ensure you have `org.eclipse.jdt.annotation` bundle on you buildpath (with addition in this PR)
2. Enter the following lines in the project's `.settings/org.eclipse.jdt.core.prefs`:
```
org.eclipse.jdt.core.compiler.annotation.resourceanalysis=enabled
org.eclipse.jdt.core.compiler.problem.potentiallyUnclosedCloseable=warning
org.eclipse.jdt.core.compiler.problem.unclosedCloseable=error
```

In particular the first option doesn't yet have UI. By specifying `org.eclipse.jdt.core.compiler.annotation.owning=your.Annot` you could optionally roll you own equivalent of `@org.eclipse.jdt.annotation.Owning`.

With this setup interested parties can play with the new analysis. Have a look at new tests or the following example for inspiration:

```
public class Producer {
	@Owning InputStream get() throws FileNotFoundException {
		return new FileInputStream("foo");
	}
}
```
```
public class Consumer {
	@Owning InputStream inspect(@Owning InputStream input) {
		if (input != null)
			System.out.println("ok");
		return input;
	}
	public void consume(@Owning AutoCloseable rc) {
		try (rc) {
			System.out.println("reading");
		} catch (Exception e) {
			e.printStackTrace();
		}
	}
}
```
```
public class Client {
	void ok(Producer p, Consumer c) throws IOException {
		InputStream is = p.get(); // *1*
		InputStream is2 = c.inspect(is); // *2* 
		c.consume(is2);  // *3*
	}
}
```
By commenting line \*3* and then \*2* you will see errors appear on the previous line. By removing the annotation from `Producer.get()` the error will disappear again. Etc. pp.
